### PR TITLE
Source-term interface

### PR DESCRIPTION
**Pull request by hyschive**
This PR implements an interface for adding new source terms.

Supported features:
- Built-in source terms (a template of deleptonization can be found at `src/SourceTerms/Deleptonization`)
- User-specified source terms (a template can be found at `src/SourceTerms/User_Template`)
- GPU acceleration

Future work:
- Support ghost zones
- Avoid transferring useless fields
- Wiki

### CODE REVIEW

File: src/Main/EvolveLevel.cpp [hunk 1]
```diff
@@ -434,7 +435,17 @@ void EvolveLevel( const int lv, const double dTime_FaLv )
 // ===============================================================================================
 
 // *********************************
-//    6-1. Grackle cooling/heating
+//    6-1. local source terms
+// *********************************
+   const int InOutSg_Src = SaveSg_Flu;    // save in the same FluSg
+
```
Comment (hfhsieh) [inline src/Main/EvolveLevel.cpp:441]: - `InOutSg_Src` is not used
- Adjust the indentation of these lines

File: src/Main/InvokeSolver.cpp [hunk 2]
```diff
@@ -96,6 +97,9 @@ void InvokeSolver( const Solver_t TSolver, const int lv, const double TimeNew, c
       Aux_Error( ERROR_INFO, "incorrect SaveSg_Flu (%d) !!\n", SaveSg_Flu );
 #  endif
 
+   if ( TSolver == SRC_SOLVER  &&  ( SaveSg_Flu != 0 &&  SaveSg_Flu != 1 )  )
+      Aux_Error( ERROR_INFO, "incorrect SaveSg_Flu (%d) !!\n", SaveSg_Flu );
+
```
Comment (hfhsieh) [inline src/Main/InvokeSolver.cpp:102]: For these checks, also print the TSolver could be helpful for debugging.

File: src/SourceTerms/User_Template/CPU_Src_User_Template.cpp [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+// Return      :  None
+//-----------------------------------------------------------------------------------------
+void Src_Init_User_Template()
+{
+
+   Src_SetAuxArray_User_Template( SRC_TERMS.User_AuxArray_Flt, SRC_TERMS.User_AuxArray_Int );
+   Src_SetFunc_User_Template( SRC_TERMS.User_FuncPtr );
+
+} // FUNCTION : Src_Init_User_Template
+
```
Comment (hfhsieh) [inline src/SourceTerms/User_Template/CPU_Src_User_Template.cpp:217]: Add `Src_End_User_Template()` here?

File: src/SourceTerms/Src_Prepare.cpp [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+//                PID0_List        : List recording the target patch indices with LocalID==0
+//-------------------------------------------------------------------------------------------------------
+void Src_Prepare( const int lv, const double PrepTime,
+                  real h_Flu_Array_S_In[][FLU_NIN_S][ CUBE(SRC_NXT)           ],
+                  real h_Mag_Array_S_In[][NCOMP_MAG][ SRC_NXT_P1*SQR(SRC_NXT) ],
+                  double h_Corner_Array_S[][3],
+                  const int NPG, const int *PID0_List )
+{
+
+
```
Comment (hfhsieh) [inline src/SourceTerms/Src_Prepare.cpp:38]: Remove one extra empty line here.

**Review by kuochuanpan — APPROVED** [review]

**Review by Inhyeok-Song — COMMENTED** [review]
All look fine to me.

**Comment by hyschive** [comment]
@kuochuanpan @hfhsieh @Inhyeok-Song Thanks for reviewing this PR! I've incorporated all your suggestions and will merge this PR right away.

**Pull request merged**